sphinx/ext/autodoc/importer.py

```diff
@@ -31,6 +31,14 @@
 class _MockObject(object):
     """Used by autodoc_mock_imports."""
 
+    def __new__(cls, *args, **kwargs):
+        # type: (Any, Any) -> Any
+        if len(args) == 3 and isinstance(args[1], tuple) and args[1][-1].__class__ is cls:
+            # subclassing MockObject
+            return type(args[0], (_MockObject,), args[2], **kwargs)  # type: ignore
+        else:
+            return super().__new__(cls)  # type: ignore
```

Comment:
not works with py2. Is `super(M, cls)` correct?
```
            return super(M, cls).__new__(cls)  # type: ignore
```

test.py
```
from __future__ import print_function

class M(object):
    def __new__(cls, *args, **kwargs):
        print('new =>', cls, args, kwargs)
        if len(args) == 3 and isinstance(args[1], tuple) and args[1][-1].__class__ is cls:
            # subclassing MockObject
            return type(args[0], (M,), args[2], **kwargs)  # type: ignore
        else:
            return super(M, cls).__new__(cls)  # type: ignore

t1 = 'm = M(1, b=2)'
print('exec:', t1); exec(t1)

print('subclassing M')
class F(M):
    def __init__(self, a, b=0):
        self.a = a
        self.b = b

t2 = 'f = F(1, b=2)'
print('exec:', t2); exec(t2)


print('subclassing m')
class X(m):
    def __init__(self, a, b=0):
        self.a = a
        self.b = b

t3 = 'x = X(1, b=2)'
print('exec:', t3); exec(t3)
```

run with py3
```
$ python3 test.py 
exec: m = M(1, b=2)
new => <class '__main__.M'> (1,) {'b': 2}
subclassing M
exec: f = F(1, b=2)
new => <class '__main__.F'> (1,) {'b': 2}
subclassing m
new => <class '__main__.M'> ('X', (<__main__.M object at 0x1013b4400>,), {'__module__': '__main__', '__qualname__': 'X', '__init__': <function X.__init__ at 0x1013a98c8>}) {}
exec: x = X(1, b=2)
new => <class '__main__.X'> (1,) {'b': 2}
```

run with py2
```
$ python2 test.py 
exec: m = M(1, b=2)
new => <class '__main__.M'> (1,) {'b': 2}
subclassing M
exec: f = F(1, b=2)
new => <class '__main__.F'> (1,) {'b': 2}
subclassing m
new => <class '__main__.M'> ('X', (<__main__.M object at 0x102132c50>,), {'__module__': '__main__', '__init__': <function __init__ at 0x102135b18>}) {}
exec: x = X(1, b=2)
new => <class '__main__.X'> (1,) {'b': 2}
```

Reply:
Right. I'd forgotten the usage of `super()` in py2...